iXBRLViewerPlugin/ui.py

```diff
@@ -72,33 +80,96 @@ def __init__(self, cntlr):
         self.grab_set()
         self.wait_window(self)
 
-   
     def close(self, event=None):
+        """
+        Closes the dialog without performing "ok" actions
+        :param event:
+        :return:
+        """
         self.parent.focus_set()
         self.destroy()
 
+    def getTitle(self) -> str:
+        pass
+
+    def addFields(self, frame: Frame, y: int) -> int:
+        """
+        Adds fields specific to the viewer dialog implemented
+        :param frame: Frame to add fields to
+        :param y: The row (y) to start adding fields at
+        :return: The row (y) after adding fields
+        """
+        pass
+
+    def ok(self, event=None):
+        """
+        Actions to perform if OK is pressed
+        :param event:
+        :return:
+        """
+        pass
+
+
+class SaveViewerDialog(ViewerDialog):
+    """
+    Dialog for saving a viewer instance.
+    Initializes with (but doesn't overwrite) default settings configured in the settings dialog.
+    """
+
+    def __init__(self, cntlr):
+
+        self.cntlr = cntlr
+        self.parent = cntlr.parent
+
+        self._filename = StringVar()
+        self._filename.set(self.cntlr.config.setdefault(CONFIG_OUTPUT_FILE, ""))
+        self._zipViewerOutput = BooleanVar()
+        self._zipViewerOutput.set(self.cntlr.config.setdefault(CONFIG_ZIP_OUTPUT, False))
+        self.accepted = False
+
+        super(SaveViewerDialog, self).__init__(cntlr, "Save iXBRL Viewer")
```

Comment:
Typically you call super before any customizations in the sub class. Does something break if you move this to the first line of the `__init__`?

Reply:
Yes, the bulk of the UI layout/logic happens in `ViewerDialog.__init__`, and it depends on these variables being initialized ahead of time. I wouldn't have done it this way if it wasn't an established pattern both here and in Arelle dialogs. I think a better solution is to move the rendering logic out of the initialization code and move it to some `render()` function.

Reply:
Thanks for the explanation 🙂